src/server.ts

```diff
@@ -13,107 +12,127 @@ declare interface Context {
+    if (Component.prototype && Component.prototype.isReactComponent) {
+      const instance = new Component(props, context);
+
+      // Override setState to just change the state, not queue up an update.
+      //   (we can't do the default React thing as we aren't mounted "properly"
+      //   however, we don't need to re-render as well only support setState in
+      //   componentWillMount, which happens *before* render).
+      instance.setState = (newState) => {
+        instance.state = assign({}, instance.state, newState);
       };
-    } catch (e) {} // tslint:disable-line
-    if (Component.componentWillMount) Component.componentWillMount();
 
-    let newContext = context;
-    if (Component.getChildContext) newContext = assign({}, context, Component.getChildContext());
+      // this is a poor man's version of
+      //   https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactCompositeComponent.js#L181
+      if (instance.componentWillMount) {
+        instance.componentWillMount();
+      }
+
+
+      if (instance.getChildContext) {
+        childContext = assign({}, context, instance.getChildContext());
+      }
 
-    // see if there is a fetch data method
-    if (typeof type.fetchData === 'function' && fetch) {
-      const query = type.fetchData(ownProps, newContext);
-      if (query) queries.push({ query, component });
+      child = instance.render();
+    } else { // just a stateless functional
+      child = Component(props, context);
     }
 
-    getQueriesFromTree({
-      component: getChildFromComponent(Component),
-      context: newContext,
-      queries,
-    });
-  } else if (props && props.children) {
-    Children.forEach(props.children, (child: any) => getQueriesFromTree({
-      component: child,
-      context,
-      queries,
-    }));
+    walkTree(child, childContext, visitor);
+
+  } else { // a basic string or dom element, just get children
+    if (element.props && element.props.children) {
+      Children.forEach(element.props.children, (child: any) => {
+        walkTree(child, context, visitor);
+      });
+    }
   }
+}
+
+function getQueriesFromTree(
+  { rootElement, rootContext = {} }: QueryTreeArgument, fetchRoot: boolean = true
+): QueryResult[] {
+  const queries = [];
+
+  walkTree(rootElement, rootContext, (element, context) => {
+    const Component = element.type || element;
+
+    const skipRoot = !fetchRoot && (element === rootElement);
+    if (typeof Component.fetchData === 'function' && !skipRoot) {
+      const props = assign({}, Component.defaultProps, element.props);
+      const query = Component.fetchData(props, context);
+      if (query) {
+        queries.push({ query, element, context });
+
+        // Tell walkTree to not recurse inside this component;  we will
+        // wait for the query to execute before attempting it.
+        return false;
```

Comment:
Ahhh, of course. Somehow I missed the fact that there are two separate recursions happening: one to find all top-level `graphql` nodes, and one to find/execute their child queries once the first have completed... although, of course, you had clearly stated that. 

> We have to wait for X's query though, as any sub-queries may rely on the results.

Of course. Presumably in the future some introspection could check whether this is true, and batch the queries together if not. 

Hah, well I'm glad to get your thoughts on the public record while we can then! 😄